hazelcast/src/test/java/com/hazelcast/cluster/TcpIpHostnameJoinTest.java

```diff
@@ -0,0 +1,140 @@
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.hazelcast.cluster;
+
+import com.hazelcast.config.Config;
+import com.hazelcast.core.Hazelcast;
+import com.hazelcast.core.HazelcastInstance;
+import com.hazelcast.instance.impl.DefaultAddressPicker;
+import com.hazelcast.instance.impl.HazelcastInstanceFactory;
+import com.hazelcast.internal.util.ExceptionUtil;
+import com.hazelcast.test.annotation.SlowTest;
+import org.junit.After;
+import org.junit.BeforeClass;
+import org.junit.Test;
+import org.junit.experimental.categories.Category;
+import org.junit.runner.RunWith;
+import org.powermock.core.classloader.annotations.PowerMockIgnore;
+import org.powermock.core.classloader.annotations.PrepareForTest;
+import org.powermock.modules.junit4.PowerMockRunner;
+
+import java.io.IOException;
+import java.net.InetAddress;
+import java.net.NetworkInterface;
+import java.net.UnknownHostException;
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.List;
+
+import static com.hazelcast.test.HazelcastTestSupport.assertClusterSize;
+import static java.util.Collections.enumeration;
+import static org.mockito.Mockito.when;
+import static org.powermock.api.mockito.PowerMockito.mock;
+import static org.powermock.api.mockito.PowerMockito.mockStatic;
+
+@RunWith(PowerMockRunner.class)
+@PrepareForTest(DefaultAddressPicker.class)
+@PowerMockIgnore("javax.management.*")
+@Category(SlowTest.class)
+public class TcpIpHostnameJoinTest {
+
+    private static final InetAddress LOCALHOST;
+    private static final InetAddress LOCALHOST2;
+    private static final InetAddress LOCALHOST3;
+    private static final InetAddress LOCALHOST4;
+
+    static {
+        try {
+            LOCALHOST = InetAddress.getByAddress("localhost", new byte[] {127, 0, 0, 1});
+            LOCALHOST2 = InetAddress.getByAddress("localhost2", new byte[] {127, 0, 0, 1});
+            LOCALHOST3 = InetAddress.getByAddress("localhost3", new byte[] {127, 0, 0, 1});
+            LOCALHOST4 = InetAddress.getByAddress("localhost4", new byte[] {127, 0, 0, 1});
+        } catch (UnknownHostException e) {
+            throw ExceptionUtil.rethrow(e);
+        }
+    }
+
+    @BeforeClass
+    public static void beforeClass() throws IOException {
+        mockStatic(NetworkInterface.class);
+
+        List<NetworkInterface> networkInterfaces = new ArrayList<NetworkInterface>();
+        networkInterfaces.add(createNetworkConfig("lo", true, LOCALHOST, LOCALHOST2, LOCALHOST3, LOCALHOST4));
+        when(NetworkInterface.getNetworkInterfaces()).thenReturn(enumeration(networkInterfaces));
```

Comment:
```suggestion
        when(NetworkInterface.getNetworkInterfaces()).then(invocation -> enumeration(networkInterfaces));
```

to return a new enumeration each time `getNetworkInterfaces()` is called.